### ISSUE TITLE
Github Pages website

### ISSUE THREAD

**Issue by onebree** [andreareginato/betterspecs#153]
In partnership with issue #146 

Previous participants: @andreareginato , @akz92

As discussed in the named issue, we want to move away from Heroku and instead host the site on Github. There are two ways of doing this.

1. Launch the site nearly as-is, HTML and all, onto Github Pages. @akz92 has already done this to create a [live demo][1].
  - Pro: Most of the HTML can remain the same
  - Con: Currently no syntax highlighting
2. Use Jekyll, a static pages framework.
  - Pro: Markdown natively supported. (As well as syntax highlighting)
  - Con: Possible restructure of the site
    + Moving all "Guidelines" to their own pages
    + Redesigning current layouts to accomodate
  - Pro: Use Disqus for conversations, rather than always-opened Github issues
  - Pro: possibly easier management of translated sites

If we proceed with straight-up HTML, we can move out of beta much quicker.

If we move to Jekyll, we can still use much of the HTML from the original site. Markdown will recognize full HTML and render it as such, skipping the md>html mapping. We can also choose from a couple of themes ready-made, but easy to edit

- [Documentation Theme][2]
  + Similar structure to current website
  + Can easily edit margins/padding in CSS
  + Native Disqus support
- [Example of BlackDoc][3]
  + I personally have experience with this theme
  + No native Disqus support
  + Optimized for documentation-heavy pages
  + Colors can be easily reverted/changed to match the current site 

[1]: https://github.com/akz92/akz92.github.io
[2]: http://idratherbewriting.com/documentation-theme-jekyll/
[3]: http://karloespiritu.com/cheatsheets/

**Comment by akz92** [comment]
@onebree my fork is using Jekyll.

**Comment by benoittgt** [comment]
First thanks a lot for the job done.

Do you have a task list of missing stuff for this new version?

**Comment by onebree** [comment]
@benoittgt not at the moment. I have not inspected the demo very much. I will add a checklist to the parent comment before next week.

**Comment by onebree** [comment]
@benoittgt I finally added the checklist to the master comment!

@andreareginato I found out that Gitbook now allows for discussion boards and in-line comments for a publication.
  - https://www.gitbook.com/blog/features/discussions
  - Unlimited _public_ projects for free (should not be an issue)
  - Provides documentation-like navigation
  - Allows for custom landing page
  - Example free gitbook: https://github.com/DjangoGirls/tutorial

I know I was extremely excited about the Jekyll site. I just wanted to know your thoughts on this new thing I found.

**Comment by franzejr** [comment]
I think Gitbook is pretty awesome and simple.

**Comment by onebree** [comment]
@franzejr yes, gitbook is good. IIRC, there were other OSS alternatives out there, but I cannot find which ones. If you do, let me know so I can review them as well. I think anything like Gitbook would allow for quicker releases, and less worrying about the design. (Which is the current hold-up)

@akz92 - I want to say thank you for all you have done porting the site so far to Jekyll. Unfortunately, I have not had time to really rummage through it or launch locally. I would like you know you and @andreareginato thoughts on Gitbook. Like I just said, it would mean less time worrying about design details and more time thinking about content.

**Comment by dankohn** [comment]
Positives of Gitbook are that it looks nice and would enable free downloads of the contents as eBooks and PDFs, which some users would I'm sure like.

Downside is that Gitbook/GitHub integration is only [one way](https://help.gitbook.com/github/index.html). When any change is made to the Github repo, it will automatically propagate to the Gitbook. And, using the GitBook editor will make changes to the Github repo.  But, these changes will not be automatically compiled into HTML usable by Github Pages to actually display the website. To update the website would require an additional step of an authorized committer re-rendering the webpages and uploading to Github. By contrast, a Jekyll back-end will automatically update Github Pages on every commit.

It does seem possible to setup a Travis post-commit hook that would automatically render the Github Pages HTML and commit them back to the repo, as [documented](http://www.steveklabnik.com/automatically_update_github_pages_with_travis_example/). But that would be some meaningful additional work.

**Comment by onebree** [comment]
@dankohn git hooks are on my list of things to learn. My comment from 11/09 suggests we would use Gitbook _instead of_ Jekyll. However, if Gitbook limits the design customization (no response from them yet), then a Github Pages site would still be handy.

If we need these git hooks, you are welcome to help us out. I would add you as a collaborator etc at that point.

**Comment by dankohn** [comment]
I believe you should choose either Gitbook or Jekyll for the back-end of the website, and should use Github Pages to actually serve the HTML. I think you could make either work. Jekyll would be much easier to setup. Gitbook would also produce PDFs and eBook formats.

**Comment by dankohn** [comment]
It does seem possible to host the site on Gitbook instead of Github Pages, using the existing custom domain, as described here: https://help.gitbook.com/platform/domains.html

My hesitancy is that Github Pages is one of the most reliable hosting platforms in the world, but for this application, it probably doesn't matter.

**Comment by onebree** [comment]
@dankohn is it possible to have Gitbook fetch only certain partials from the Jekyll site? Such partials would be just markdown, and easy to translate to both formats. Maybe even have Gitbook point to a particular branch, if possible.

**Comment by dankohn** [comment]
Anything is possible, but Jekyll and Gitbook do not use identical flavors of Markdown, so this is harder than it looks. I strongly recommend picking one or the other as the back-end format.

**Comment by onebree** [comment]
@dankohn thank you for your insight. I brought Gitbook up because I am personally moving away from Jekyll (with my blog), and I like the documentation-style of it.

In regards to Markdown, I know you can specify the flavor in Jekyll, such as GFM (which Gitbook supports).

If, at the end of the day, Gitbook only provides the PDF/ebook feature, we could find an alternative route within Jekyll (or a different generator altogether). Basically, the end-user can always view the Markdown generated file in Github, and print off from there. The issue with that is Liquid syntax (which powers Jekyll) does not render nicely on Github. (It's mainly used for syntax highlighting.)

**Comment by akz92** [comment]
@onebree sorry for the really long delay to answer. I had never seen GitBook, I visited a few pages and it looks good for documentation purpose but personally I believe it isn't the best alternative for betterspecs. Using GitBook we won't need to worry about design but betterspecs design is done, I don't think it will be a problem. On top of that, the Jekyll version is almost done and it solves the main problem of the project.

**Comment by andreareginato** [comment]
Hi everyone. Next week (probably on Monday) I'll take one day off to move betterspecs.org to Github pages and to give it a new restyle. This should make changes easier for the community.

I would love to ask @onebree (which is greatly maintaining the site) to give me all the info I need, not missing any update (for what can be possible).

Thanks everyone.

**Comment by onebree** [comment]
@andreareginato Right now, @akz92 has a demo of the site in Jekyll. I have not looked at it much, however. An alternative to Jekyll/GH Pages is this, which can _easily_ be styled with the current color scheme. 

https://github.com/jeromegn/DocumentUp

It has a GH pages guide, and the [`v1` theme] looks exactly like what we want -- a sidebar with headings/links. I even found how to remove the "Fork me on Github" banner from the final site.

Once the site is ported over to GH pages, information needs to be updated for Rails 4 and RSpec 3. This is greatly important as Rails 5 is in beta.

[1]: https://github.com/jeromegn/DocumentUp/blob/master/app/assets/images/v1.png

**Comment by akz92** [comment]
@andreareginato feel free to use any code of my port. I'll be glad to help if there's something I can do.

**Comment by onebree** [comment]
@andreareginato if you do use @akz92's port, I think it would be best to push his commits to a temp branch on github. Then he will still get credit in the commit history for his work :-)

**Comment by andreareginato** [comment]
Hi everyone. 

A [draft for the new version of Better Specs](http://lelylan.github.io/betterspecs) is up and running.

### New features.
- [x] Works on Github pages (this will simplify the contribution process)
- [x] Responsive (works well on mobile and tablet)
- [x] Cleaned up the whole content
- [x] New layout (more 2016)

I would love to ask @onebree, @akz92 (thanks for the initial porting), @dankohn, @benoittgt and all better specs community to take some time, if possible, to close the last needed task before going online with the new version, which regards to the translations.
- [ ] Adapt all "translation pages" to the new layout (this will take a while, but it should not bring difficulties)

There is also nice to have todo list (see below) which is not mandatory for the release and that we can talk about later, once released on Github pages.

- [ ] Review all guidelines (they have been written more than 3 years ago and it needs a deep polish)
- [ ] Place all translation pages in a more structured way, while maintaining todays' paths
- [ ] Better use of Markdown (if possible)
- [ ] How do we handle comments? Which could be other tools to use in order to replace Github issues.
- [ ] When reloading the page with an anchor, the correct section is not loaded

Let me know what do you think guys, and thanks a lot for your time.

**Comment by onebree** [comment]
## Overall

@andreareginato  I absolutely love what you've done! I think the sidebar can remain dark, as well as code samples, but I would like to see a light page background. It adds some contrast, and is easier on the eye (all dark parts are blocked out then, not blending into the background).

## Commenting

- Disqus
  + Requires a separate account, or OAuth through Facebook/Google/Twitter
  + No code formatting (that I am aware of)
  + Comments will be tied to the page, however, not to the guideline (like how you see on blogs)
- Discourse
  + Forum-format with threading, similar to how we are doing it on Github Issues
  + Can OAuth with a Github account (which clearly everyone has already)
  + Written in Ruby and is open source
- Gitter
  + Real-time discussion of issues
  + Sign in with Github account (only option?)
  + No invite required to join
- Google Groups
  + Most everyone has a Google account, or has experienced gGroups
  + Threaded style, allowing one post per guideline

_Personally_, I would go with Google Groups or Discourse. Either way, I can help push all existing issues to the new platform, and pretty much summarize what everyone has said.

## Markdown

I initially suggested Jekyll, and later DocumentUp, because both seemed to have Markdown support out-of-the-box (even Github flavored markdown). But after a little research, I found [nanoc's guide to Markdown][1], which looks very helpful. We would just title the pages `Index`, and go on to have Markdown fun.

One issue is that Rouge supports GFM, which is why Github Pages is moving to using that gem. ([See blog post.][2]) I do not think the issue will be hard to handle, but definitely something to keep in mind.

On a side note, how  about moving to HAML? I mean, who enjoys looking at regular HTML these days?

[1]: http://nanoc.ws/doc/tutorial/#write-pages-in-markdown
[2]: https://github.com/blog/2100-github-pages-now-faster-and-simpler-with-jekyll-3-0

**Comment by dankohn** [comment]
Congrats, I think the new pages look great. I would prefer to see discussion just use Github issues, as it is the easiest and most popular option.

**Comment by onebree** [comment]
@dankohn what says it is the most popular option? I mean, it was done with this repo because I assume other tools were not available, or useful, at the time of creation.

**Comment by dankohn** [comment]
@onebree Most popular in the sense that nearly all developers have a Github account, while far less have Disqus, Discourse, Gitter, etc. Note that I'm not arguing that Github issues are the most feature-rich or even necessarily the best option, just the most ubiquitous.

**Comment by onebree** [comment]
@dankohn Well you can sign into Discourse and Gitter with Github. In fact, Github is the only login option for Gitter, I think!

**Comment by andreareginato** [comment]
Hi guys. Thanks a lot for your answers. 

If you guys agree I would keep this discussion to close the issues that are important to go online with the new version as soon as we can. To make this happen the issues to close are these.

- [ ] Adapt all "translation pages" to the new layout (this will take some hours, but it should not be difficult)
- [ ] (BUG) When reloading the page with the anchor the page doesn't go to the right section.

We need to give a priority. When we close these issues we can move to the tech ideas you were talking about. I know it's "the dirty stuff", but this is what we need to do to move to Github pages and let anyone partecipate.

**Comment by onebree** [comment]
@andreareginato What browser were you using? On Firefox 44, I had no issue with the anchors. (That is `betterspecs.org/index.html#describe`, right?)

**Comment by andreareginato** [comment]
Looking at the new version. The link below sometimes it stay at the top of the page.
http://lelylan.github.io/betterspecs/#mock

I'm using Chrome 47.0.2526.111

**Comment by onebree** [comment]
I get that issue as well in Chrome if I `CTRL+F5`. But if I visit that link directly, I am brought to the correct heading.

**Comment by onebree** [comment]
@andreareginato I just whipped this Markdown file up. It is not perfect, and has to be rendered in a test environment, but this is what the English index could look like:

https://gist.github.com/onebree/0147c9dc28c9a41d81f3

**Comment by akz92** [comment]
@andreareginato we could use an I18n translation like this: https://github.com/screeninteraction/jekyll-multiple-languages-plugin
It would make it a lot easier to maintain translations.

**Comment by andreareginato** [comment]
@akz92 looks a great solution, we can go for it.
@onebree at first looks great (it simplifies reading and writing). I would bring attention to the css structure, as we need to keep it coherent with actual HTML.

Thanks a lot guys

**Comment by andreareginato** [comment]
* @akz92 looks a great solution, we can go for it.
* @onebree at first looks great (it simplifies reading and writing). I would bring attention to the css structure, as we need to keep it coherent with actual HTML. I also should note the the actual index page (in gh-pages branch) has some consistent changes on the menu.

Let me know your thoughts

**Comment by andreareginato** [comment]
One more thing. 

For one week I'll be out for work and I'll not be able to closely follow the progress of the works. 

I wanted to leave you guys the task of making betterspecs 2.0 ready to go online telling you one more thing. Although many tech solutions can be applied, I would suggest to keep those minimal. I think going online with bs 2.0 using the `internationalization plugin` and the `markdown syntax` (where possible), would be a great win. Of course, these thought are made to shorten the time to go online with the new version as soon as we can, and many other improvements can be made after this.

Good work to everyone and thanks

**Comment by onebree** [comment]
@andreareginato The CSS was not maintained in the gist I sent you, as it did not have any accompanying CSS files. This is why I need to test this locally with the entire site. I will make a separate branch called "markdown" or something. I also just realized that the gist I made was based on master branch, not `gh-pages`.

Would you be opposed to simultaneously moving the site to HAML? This I can easily do myself, and I believe HAML (possibly with a gem) allows for Markdown syntax. (Similar to it allowing JS with the tag `:javascript`.) If you want to wait on this, I am okay with that -- it was just a suggestion to get two birds with one stone, and cleaning up the HTML a lot.

Nevertheless, I will be able to work on the Markdown structuring this weekend, as I am off from work Monday (15th).

Something I can do Friday/Saturday (which will be quick) is cleaning up the README and finishing the changelog.

**Comment by andreareginato** [comment]
@onebree I 'm not against HAML but we need to be sure to close this issue. Do you think using HAML will add much time on the process? If you can take care of it, and the final result is a much cleaner result, I'm in. Get in sync with @akz92 about the translations and I think you are ready to start. 

Great work. Lets make it real this weekend :)

**Comment by onebree** [comment]
@andreareginato is the slack channel still around? I know I did not delete it. If possible, I can add @akz92 to it, and remove anyone who has not been around.

**Comment by andreareginato** [comment]
Feel free to use it. It's up and running

Il giorno gio 11 feb 2016, 15:32 Hunter Stevens <notifications@github.com>
ha scritto:

**Comment by onebree** [comment]
@akz92 Would you like to join the Slack channel? Right now, it just posts updates from github. If you are interested, please let me know which email to use.

Furthermore, I found this comment, where someone suggested using YAML for translations (eliotsykes):

https://github.com/lelylan/betterspecs/issues/150#issuecomment-125200165

> Just in case its of interest and not been discussed, putting the translations into YAML and adding betterspecs to the public projects here https://github.com/Locale (https://www.localeapp.com/) might gain some translators/maintainers.

**Comment by onebree** [comment]
@andreareginato I have updated the changelog for all changes since I became a contributor: 45cbdb6 

Happy 300 commits!

**Comment by akz92** [comment]
Sure @onebree, my slack account email is aride.moulin@gmail.com. 
The translation solution I proposed uses YAML too. The kind of translation service mentioned by @eliotsykes is great but also expensive. It has a 500 word limit and supports only 2 languages on the free plan. Besides that, I believe it's intended to be used in rails projects, I don't know if it's possible to use it in a Jekyll project.

**Comment by andreareginato** [comment]
@onebree Impressive job :) Really really nice!
@onebree about i18n although I didn't go in deep I would set as tool the one suggested by @akz92. I would start using it as soon as possible.

As soon as you can let me know how the porting is moving. I'm looking forward to tell the community the work you are doing. Have a nice beginning of week.

**Comment by benoittgt** [comment]
Sorry was really busy. For slack here is my email benoitjerk [at] hopsandfork dot com. 

Will see on which task I can work on. Maybe : *Adapt all "translation pages" to the new layout (this will take some hours, but it should not be difficult* ?

**Comment by onebree** [comment]
@akz92 and @benoittgt I will add you two to Slack today.

@andreareginato thank you, and you are welcome. I did not get to all I wanted to this holiday weekend -- I am still figuring out the Markdown port of the site. I will worry about that before HAML. I will finish the port by the weekend. :-)

**Comment by andreareginato** [comment]
Thanks for the great work and coordination. Looking forward :+1:

**Comment by pirj** [comment]
It's been two years, and [gh-pages](https://github.com/lelylan/betterspecs/commits/gh-pages?before=0828fa7a2600f4254b37976ddc0425eee1f4f7a8+35) that accepts all pull request has diverged from master that is deployed to Heroku that makes the latter massively outdated.

Examples:
1) [pull](https://github.com/lelylan/betterspecs/commit/0d8c44730d97ad4d340c83f26a28dfd27a9c7bfe) [betterspecs.org](http://www.betterspecs.org/#data) [gh page](http://lelylan.github.io/betterspecs/#data)
2) [pull](https://github.com/lelylan/betterspecs/pull/157/files) [betterspecs.org](http://www.betterspecs.org/#contexts) [gh page](http://lelylan.github.io/betterspecs/#contexts)

This is a massive change, and it didn't work out to make it in one go.

Please mind switching `betterspecs.org` to point to GH, split this ticket into smaller chunks and work on them gradually.

**Comment by pirj** [comment]
Hi again,

[RSpec style guide](https://github.com/rubocop-hq/rspec-style-guide) has recently been moved to [RuboCop HQ]([RSpec style guide](https://github.com/rubocop-hq), and is actively maintained.

I was thinking about [cherry-picking the practices from this great guide](https://github.com/rubocop-hq/rspec-style-guide/issues/26) (both `master` and `gh-pages`, but English only) to RSpec style guide, keeping all the attribution and credits where possible.

My only concern is the legal aspect of doing so since the license is not defined for this project.

Appreciate any feedback, and glad to discuss how to join forces on creating a better comprehensive guide that the community would benefit from.

/cc @FedeG @andreareginato @benoittgt @akz92

**Comment by benoittgt** [comment]
I think it's a very good idea :)

**Comment by pirj** [comment]
I'm not sure where it's better to ask, so I'll leave the reference to the pull request to RSpec style guide picking the rules from BetterSpecs here: https://github.com/rubocop-hq/rspec-style-guide/pull/55
It has definitely made the style guide better!

**Comment by franklinyu** [comment]
So what’s the progress? Anything I can help?

**Comment by benoittgt** [comment]
@franklinyu it was merged. https://github.com/rubocop-hq/rspec-style-guide/pull/55 🎉

**Comment by pirj** [comment]
@franklinyu The two guides always complemented each other and even intersected a bit.
Unfortunately, the maintenance was not active on either, but it's improving after migration of RSG to RuboCop HQ.

It would be easier for everyone to maintain a single guide, and even though Bettrspecs is significantly more popular (stargazed), I've started off RSG, and preferred to apply the effort on updating it, including picking *all* the rules from Betterspecs.

Translations were lost in transition, I hope volunteers eventually restore this miss, as it is with [Ruby Style Guide](https://github.com/rubocop-hq/ruby-style-guide#the-ruby-style-guide), which is additionally available in ten languages.

The decision on how to proceed with this project is up to Betterspecs maintainers, whether to link to  RSG and add a notice that Betterspecs are not actively maintained and migrated/superseeded/evolutioned to RSG, or to keep updating it, specifically by backporting the rules I've updated to the new realms from RSG.

Betterspecs remains a great resource, and 95% of it still applies, it has useful links to books and presentations, and again, is translated into different languages.

No matter how it goes I'd like to thank all Betterspecs contributors and maintainers for the hard work. ❤️
